gradle/generation/javacc.gradle

```diff
@@ -0,0 +1,102 @@
+// Add a top-level pseudo-task to which we will attach individual regenerate tasks.
+import static groovy.io.FileType.*
+
+configure(rootProject) {
+  configurations {
+    javacc
+  }
+
+  dependencies {
+    javacc "net.java.dev.javacc:javacc:${scriptDepVersions['javacc']}"
+  }
+
+  task javacc() {
+    description "Regenerate sources for corresponding javacc grammar files."
+    group "generation"
+
+    dependsOn ":lucene:queryparser:javaccParserClassic"
+    dependsOn ":lucene:queryparser:javaccParserSurround"
+    dependsOn ":lucene:queryparser:javaccParserFlexible"
+  }
+}
+
+// We always regenerate, no need to declare outputs.
+class JavaCCTask extends DefaultTask {
+  @Input
+  File javaccFile
+
+  JavaCCTask() {
+    dependsOn(project.rootProject.configurations.javacc)
+  }
+
+  @TaskAction
+  def generate() {
+    if (!javaccFile || !javaccFile.exists()) {
+      throw new RuntimeException("JavaCC input file does not exist: ${javaccFile}")
+    }
+    // Remove old files so we can regenerate them
+    def parentDir = javaccFile.parentFile
+    parentDir.eachFileMatch FILES, ~/.*\.java/, { file ->
```

Comment:
If these files get overwritten I don't think we should care about explicit deletions (?).

Reply:
Oh, what's FILES -- I don't know this construct.

Reply:
I copied it from some example and it worked.......

Reply:
Actually, they aren't overwritten. If they're not deleted you get messages during execution like: "Warning: TokenMgrError.java: File is obsolete.  Please rename or delete this file so that a new one can be generated for you."

Reply:
Leaving this in. We could make it conditional on whether there's anything to do, but that's waaaay more work than the benefit warrants for the 5 seconds this takes every two years when we actually need to regenerate these.